backup: report when job is slow due to overload

dt opened the pull request:
Previously when looking at a job making less progress than normal, it was sometimes difficult to determine if it was being slowed due to admission control protecting the cluster from overload. One _can_ go to the overload dashboard and see if _something_ is being slowed due to AC by looking at the admission delay micros/sec charts, but this just tells you that _something_ or _somethings_ are being slowed, not how much specifically the job you were looking at is being slowed in particular. Additionally needing to navigate to the overload dashboard to find out if that is or is not why a specific job is hanging is a step that is not intuitive. Instead, if a single (backup) job detect that its work is being substantially delayed due to admission delays in its recorded traces, it can directly surface this fact and inform the user that it in particular was delayed by a certain amount, making it quicker and easier for an operator that is looking at a slow/stuck backup job to determine what is wrong if it is stuck due to the cluster being overloaded.

Release note: none.
Epic: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142352)
<!-- Reviewable:end -->

dt committed:
util/bulk: add a trace-backed 'delay logger' helper

Release note: none.
Epic: none.

dt committed:
backup: detect and log overload delays

When a backup is running more slowly because admission control is delaying its
progress, it can be easy to miss: one needs to look at specifically the overload
dashboard and even then how much a specific backup or changefeed or other job is
specifically being impacted is not very clear.

This change extends backup to use its collected trace infomation, which includes
information about how long the backup spent doing work/waiting for work to be done
as well as how long the backup's work spent delayed due to AC, and then logs a warning
to the backup job's message log if that ratio exceeds 50%.

Release note: none.
Epic: none.

dt commented:
@msbutler @jeffswenson I tried to refactor this into something that should be easy to replicate in other jobs; what do you think of the new structure?

dt committed:
backup: add trace span per worker

Release note: none.
Epic: none.

dt committed:
util/bulk: dump trace stats in stable order

Randomized map iteration makes it harder to compare two dumps side by side
since they are presented in different orders. There isn't any logical relationship
between span names to preserve, so just alphabetizing them is as good as any
other stable ordering.

Release note: none.
Epic: none.

dt commented:
Probably want to dig into https://github.com/cockroachdb/cockroach/issues/143920 before advancing this
